client/js/widgets/scoreboard.js

```diff
@@ -0,0 +1,287 @@
+    const v = asArray(values);
+    tr.innerHTML = Array(values.length).fill(`<${cellType}></${cellType}>`).join('');
+    for (let i=0; i < values.length; i++)
+      $a(cellType, tr)[i].innerText = values[i];
+    return tr;
+  }
+
+  tableCreate(seats) {
+    /* This routine creates the HTML table for display in the scoreboard. It is
+     * complicated by the fact that the `seats` property can be either an array of
+     * seat IDs or an object whose keys are team names and each of whose values is an
+     * array of seat IDs.
+     * There are two major sections: the first computes the pScores array, which contains
+     * the array of scores, either for seats or for teams. The second section uses the
+     * pScores array to construct the HTML table.
+     * There are lots of other things going on, to get the totals line, round names, etc
+     * correct.
+     */
+    const showTotals = this.get('showTotals');
+    const scoreProperty = this.get('scoreProperty');
+    let sortField = this.get('sortField');
+    let usePlayerColors = this.get('usePlayerColors');
+    const colors = []; // Array of player colors
+
+    // Compute number of scoring rounds to show
+    let rounds = this.get('rounds'); // User-supplied round names
+    let includedSeats = seats; // Array of all relevant seats
+    if(typeof(seats)=='object') // Array of all seats for team use
+      includedSeats = Object.keys(seats).reduce((union,key) => union.concat(seats[key]), []);
+
+    let numRounds=0;
+    for (let i=0; i < includedSeats.length; i++) {
+      const score = includedSeats[i].get(scoreProperty);
+      if(Array.isArray(score) && score.length > numRounds)
+        numRounds = score.length;
+    }
+    if(this.get('showAllRounds') && Array.isArray(rounds))
+      numRounds = Math.max(rounds.length, numRounds);
+
+    let pScores = []; // Array of scores. pScores[i][0] is player name or team name, last is total (or last score if showTotals is false)
+    if(Array.isArray(seats)) { // Show individual seats
+      // Fill player score array, totals array
+      for (let i=0; i < seats.length; i++) {
+        const score = seats[i].get(scoreProperty);
+        let total;
+        if(Array.isArray(score)) {
+          pScores[i] = score.map( s => isNaN(parseFloat(s)) ? 0 : parseFloat(s));
+          total = this.getTotal(pScores[i]);
+        } else {
+          pScores[i] = [];
+          total = parseFloat(score);
+          if(isNaN(total))
+            total = 0;
+        }
+        pScores[i] = pScores[i].concat(Array(numRounds).fill('')).slice(0,numRounds);
+        if(showTotals)
+          pScores[i].push(total);
+        pScores[i].unshift(seats[i].get('player'));
+      }
+
+      // Sort player scores as requested
+      if(sortField == 'total' && !showTotals) // Use default sort if no totals
+        sortField = 'index';
+      if(sortField == 'total')
+        pScores.sort((a,b) =>a[a.length-1] < b[b.length-1] ? -1 : 1)
```

Comment:
These are numbers, right? You usually do `b[b.length-1] - a[a.length-1]` so equal values return `0` as expected by `sort`.

At least that's what I think you're supposed to do.

Reply:
The mozilla and w3c `sort` pages say different things. According to mozilla, the sort function should return 1, -1, or 0; w3c says it should be positive, negative, or zero. In any case, my code is wrong because it does need to return 0 if the two are equal. I'll change something here and below.

Reply:
Fixed.